src/cpp/src/scheduler.hpp

```diff
@@ -45,6 +52,23 @@ class Scheduler {
 
     Output schedule(std::vector<SequenceGroup::Ptr>& sequence_groups) {
         Output scheduler_output;
+        float eps = 1e-5;
+
+        if (!m_block_manager.block_allocator_initialized()) {
+            size_t prompt_sum_size = 0;
+            for (auto idx = 0; idx < sequence_groups.size(); idx++) {
+                prompt_sum_size += sequence_groups[idx]->get_prompt_len();
+            }
+            size_t initial_kv_cache_size = prompt_sum_size * m_kv_blocks_initial_multiplier;
+            m_block_manager.increase_kv_blocks_number(initial_kv_cache_size);
+            m_dynamic_memory_allocation = true;
+        }
+        else if (m_dynamic_memory_allocation && (m_block_manager.get_used_percentage() + eps) > m_precentage_threshold_for_cache_increase) {
+            size_t new_cache_size = (size_t)(m_block_manager.get_total_number_of_kv_blocks() * m_cache_growth_factor);
```

Comment:
should we also take into account device HW capabilities? E.g. on GPU we can easily be out of memory and probably need to take this into account.

E.g. we can bound new_cache_size with OpenVINO device available free memory. If on certain iteration we cannot allocate more blocks, we need to throw an exception.

Reply:
Secondly, should we take into account max_new_tokens / max_length / get_max_new_tokens ?
E.g. if we have generated 256 tokens, while max_new_tokens is 300, we don't need to multiply by 2x

BTW, do plugins have logic to decrease memory if we have allocated quite huge size on certain `generate` call?

Reply:
> should we also take into account device HW capabilities? E.g. on GPU we can easily be out of memory and probably need to take this into account.

Implemented this logic for GPU. For CPU there are no similar properties, which allow to check available memory, and CPU implementation of dynamic KV-cache allocation doesn't have logic for checking of available memory, so I implemented this only for GPU. 

> Secondly, should we take into account max_new_tokens / max_length / get_max_new_tokens ?

I implemented this for initializing of cache. Regarding cache increasing, this part was moved to scheduler and as discussed implemented simple multiplication by two there. 

> BTW, do plugins have logic to decrease memory if we have allocated quite huge size on certain generate call?

Plugins do not have this logic. As we discussed, didn't add this for now.